vcd/resource_vcd_vm_internal_disk.go

```diff
@@ -0,0 +1,475 @@
+
+	diskSetting := &types.DiskSettings{
+		SizeMb:              int64(d.Get("size_in_mb").(int)),
+		UnitNumber:          d.Get("unit_number").(int),
+		BusNumber:           d.Get("bus_number").(int),
+		AdapterType:         internalDiskBusTypes[d.Get("bus_type").(string)],
+		ThinProvisioned:     &isThinProvisioned,
+		StorageProfile:      storageProfilePrt,
+		Iops:                &iops,
+		VirtualQuantityUnit: "byte",
+		OverrideVmDefault:   overrideVmDefault,
+	}
+
+	err = powerOffIfNeeded(d, vm)
+	if err != nil {
+		return err
+	}
+
+	diskId, err := vm.AddInternalDisk(diskSetting)
+	if err != nil {
+		return fmt.Errorf("error updating VM disks: %s", err)
+	}
+
+	d.SetId(diskId)
+
+	err = powerOnIfNeeded(d, vm)
+	if err != nil {
+		return err
+	}
+
+	return resourceVmInternalDiskRead(d, meta)
+}
+
+func powerOnIfNeeded(d *schema.ResourceData, vm *govcd.VM) error {
+	vmStatus, err := vm.GetStatus()
+	if err != nil {
+		return fmt.Errorf("error getting VM status before ensuring it is powered on: %s", err)
+	}
+
+	if (vmStatus != "POWERED_ON" && d.Get("bus_type").(string) == "ide" && d.Get("allow_vm_reboot").(bool)) ||
+		(vmStatus != "POWERED_ON" && d.Get("allow_vm_reboot").(bool) && (d.HasChange("bus_number") || d.HasChange("unit_number"))) {
+		log.Printf("[DEBUG] Powering on VM %s after adding internal disk.", vm.VM.Name)
+
+		task, err := vm.PowerOn()
+		if err != nil {
+			return fmt.Errorf("error powering on VM for adding/updating internal disk: %s", err)
+		}
+		err = task.WaitTaskCompletion()
+		if err != nil {
+			return fmt.Errorf(errorCompletingTask, err)
+		}
+	}
+	return nil
+}
+
+func powerOffIfNeeded(d *schema.ResourceData, vm *govcd.VM) error {
+	vmStatus, err := vm.GetStatus()
+	if err != nil {
+		return fmt.Errorf("error getting VM status before ensuring it is powered on: %s", err)
+	}
+
+	if (vmStatus != "POWERED_OFF" && d.Get("bus_type").(string) == "ide" && d.Get("allow_vm_reboot").(bool)) ||
+		(vmStatus != "POWERED_OFF" && d.Get("allow_vm_reboot").(bool) && (d.HasChange("bus_number") || d.HasChange("unit_number"))) {
+		log.Printf("[DEBUG] Powering off VM %s for adding/updating internal disk.", vm.VM.Name)
+
+		task, err := vm.PowerOff()
+		if err != nil {
+			return fmt.Errorf("error powering off VM for adding internal disk: %s", err)
+		}
+		err = task.WaitTaskCompletion()
+		if err != nil {
+			return fmt.Errorf(errorCompletingTask, err)
+		}
+	}
+	return nil
+}
+
+// Deletes disk from VM
+func resourceVmInternalDiskDelete(d *schema.ResourceData, m interface{}) error {
+	vcdClient := m.(*VCDClient)
+
+	vcdClient.lockParentVm(d)
+	defer vcdClient.unLockParentVm(d)
+
+	vm, _, err := getVm(vcdClient, d)
+	if err != nil {
+		return err
+	}
+
+	err = powerOffIfNeeded(d, vm)
+	if err != nil {
+		return err
+	}
+
+	err = vm.DeleteInternalDiskById(d.Id())
+	if err != nil {
+		return fmt.Errorf("[Error] failed to delete internal disk: %s", err)
+	}
+
+	err = powerOnIfNeeded(d, vm)
```

Comment:
I think it will always be powered on - even if it was powered off before. We need to check against a real state of VM.